tao_compiler/mlir/disc/transforms/dot_rewriter.cc

```diff
@@ -190,6 +191,469 @@ struct TransposeFoldingConvert : public OpRewritePattern<mhlo::DotGeneralOp> {
+      reordered_contracting_tokens.push_back(t);
+    }
+  }
+  for (char t : lhs_original_tokens) {
+    if (lhs_non_contracting_tokens.contains(t)) {
+      reordered_lhs_non_contracting_tokens.push_back(t);
+    }
+  }
+  for (char t : rhs_original_tokens) {
+    if (rhs_non_contracting_tokens.contains(t)) {
+      reordered_rhs_non_contracting_tokens.push_back(t);
+    }
+  }
+
+  // lhs/rhs
+  std::copy(reordered_batching_tokens.begin(), reordered_batching_tokens.end(),
+            std::back_inserter(lhs_reordered_tokens));
+  std::copy(reordered_batching_tokens.begin(), reordered_batching_tokens.end(),
+            std::back_inserter(rhs_reordered_tokens));
+  if (isLhsBNC) {
+    std::copy(reordered_lhs_non_contracting_tokens.begin(),
+              reordered_lhs_non_contracting_tokens.end(),
+              std::back_inserter(lhs_reordered_tokens));
+    std::copy(reordered_contracting_tokens.begin(),
+              reordered_contracting_tokens.end(),
+              std::back_inserter(lhs_reordered_tokens));
+  } else {
+    std::copy(reordered_contracting_tokens.begin(),
+              reordered_contracting_tokens.end(),
+              std::back_inserter(lhs_reordered_tokens));
+    std::copy(reordered_lhs_non_contracting_tokens.begin(),
+              reordered_lhs_non_contracting_tokens.end(),
+              std::back_inserter(lhs_reordered_tokens));
+  }
+  if (isRhsBNC) {
+    std::copy(reordered_rhs_non_contracting_tokens.begin(),
+              reordered_rhs_non_contracting_tokens.end(),
+              std::back_inserter(rhs_reordered_tokens));
+    std::copy(reordered_contracting_tokens.begin(),
+              reordered_contracting_tokens.end(),
+              std::back_inserter(rhs_reordered_tokens));
+  } else {
+    std::copy(reordered_contracting_tokens.begin(),
+              reordered_contracting_tokens.end(),
+              std::back_inserter(rhs_reordered_tokens));
+    std::copy(reordered_rhs_non_contracting_tokens.begin(),
+              reordered_rhs_non_contracting_tokens.end(),
+              std::back_inserter(rhs_reordered_tokens));
+  }
+  // result
+  std::copy(reordered_batching_tokens.begin(), reordered_batching_tokens.end(),
+            std::back_inserter(result_reordered_tokens));
+  std::copy(reordered_lhs_non_contracting_tokens.begin(),
+            reordered_lhs_non_contracting_tokens.end(),
+            std::back_inserter(result_reordered_tokens));
+  std::copy(reordered_rhs_non_contracting_tokens.begin(),
+            reordered_rhs_non_contracting_tokens.end(),
+            std::back_inserter(result_reordered_tokens));
+}
+
+Value EinsumToDotGeneralPattern::processOperand(
+    PatternRewriter& rewriter, Value original_operand, Location loc,
+    SmallVector<char> original_tokens, SmallVector<char> reordered_tokens,
+    llvm::SmallDenseSet<char> non_contracting_tokens,
+    llvm::SmallDenseSet<char> contracting_tokens,
+    llvm::SmallDenseSet<char> batching_tokens, bool isBNC) const {
```

Comment:
nit: `const Container& xxx` to avoid copy.

Reply:
done